R/buffer.R

```diff
@@ -0,0 +1,76 @@
+#' Apply an inclusion radius and exclusion buffer to indices
+#'
+#' @param data An object of class `sf` or `sfc`.
+#' @param indices List of indices in each fold generated by `split_unnamed`.
+#' @param radius Numeric: points within this distance of the initially-selected
+#' test points will be assigned to the assessment set.
+#' @param buffer Numeric: points within this distance of the test set
+#' (after `radius` is applied)
+#' will be assigned to neither the analysis or assessment set.
+#'
+#' @keywords internal
+buffer_indices <- function(data, indices, radius, buffer) {
+
+  if (sf::st_is_longlat(data) && !sf::sf_use_s2()) {
+    rlang::abort(
+      c(
+        "`buffer` and `radius` can only be used with geographic coordinates when using the s2 geometry library",
+        "i" = "Reproject your data into a projected coordinate reference system using `sf::st_transform()`",
+        "i" = "Or install the `s2` package and enable it using `sf::sf_use_s2(TRUE)`"
+      )
+    )
+  }
+
+  n <- nrow(data)
+
+  if (is.null(radius)) radius <- 0
+  if (is.null(buffer)) buffer <- 0
+
+  if (radius) {
+    indices <- row_ids_within_dist(data, indices, radius)
+  }
+
+  # `buffer_indices` are _always_ needed
+  # so initialize to integer(0) before checking if we need to find them:
+  buffer_indices <- lapply(seq_along(indices), function(x) integer(0))
+  if (buffer) {
+    buffer_indices <- row_ids_within_dist(data, indices, buffer)
+  }
+
+  purrr::map2(indices, buffer_indices, buffered_complement, n = n)
+}
+
+buffered_complement <- function(ind, buff_ind, n) {
+  list(
+    analysis = setdiff(1:n, c(ind, buff_ind)),
+    assessment = unique(ind)
+  )
+}
+
+row_ids_within_dist <- function(data, indices, dist) {
```

Comment:
This function is extremely similar to `row_ids_intersect_polygon` in `spatial_block_cv.R`, except it uses `st_is_within_distance` rather than `st_intersects`. It might be nice to combine the two, but I thought that was a bit much for right now.